### ISSUE TITLE
Train-ourselves parity protocol for converted layout generators

### ISSUE THREAD

**Issue by creative-graphic-design-dev** [creative-graphic-design/design-generators#148]
# Unified Spec: v0.5 Training Parity And LightningCLI

Chair output for Round 2 across `train-parity`, `lightning-cli`, and `layoutflow-vendor`.

## Scope

This spec starts the v0.5 train-ourselves lane. The first target is LayoutFlow: add package-local PyTorch Lightning training driven by LightningCLI YAML, then verify that the converted training loop agrees with the vendor Lightning/Hydra loop at step level before using it for retraining.

Repository writes, commits, PRs, and GitHub issue creation were not performed in this planning round. This spec assumes new issues will be created before implementation:

- v0.5 umbrella meta issue: `Train-ourselves parity protocol for converted layout generators`
- LayoutFlow implementation issue: `[layout-flow] Add LightningCLI training and vendor step parity`

Issue #21 is unrelated; it is PosterLLaVA and must not be referenced for LayoutFlow training.

## Agreements

1. Converted training uses PyTorch Lightning through the root `training` extra and LightningCLI. Hydra is vendor-only and must not be imported by converted training code.
2. Each train-ourselves model owns its `LightningModule`, `LightningDataModule`, and `configs/*.yaml` inside the model package.
3. LayoutFlow training wraps the existing `LayoutFlowTransformerModel`, preserving one inference/training state-dict surface.
4. Public inference outputs stay unchanged: trained weights are converted back into the existing `LayoutFlowPipeline` and smoke-tested with `from_pretrained`.
5. Configs are fully resolved LightningCLI YAML files using `class_path` and `init_args`; no Hydra `_target_`, `defaults`, interpolation, or `hydra.*`.
6. Vendor-compatible configs and strict deterministic parity configs are separate. Vendor-compatible mode records original seed behavior; strict mode adds deterministic controls for equality checks.
7. Step parity is staged: static config/state, fixed-batch pre-optimizer trace, one optimizer step, N fixed batches, deterministic DataLoader stream, then vendor-default statistical smoke.
8. Shared training/parity helpers live in a new independent workspace member `lib/traingen` (import package `traingen`), not under `laygen.common`; model-specific vendor trace adapters stay in each model package or `tests/vendor_parity`.
9. Add a `training` pytest marker. Real LightningCLI fit runs and vendor training parity stay outside regular CI, while fast config/datamodule/module tests remain unmarked for coverage.

## Shared Training Library

Final name: `traingen` (`lib/traingen`, import `traingen`).

Rationale:

- It matches the existing contraction style of `laygen` (layout generation) and
  `posgen` (poster generation).
- It is a valid Python import name, short, and describes training support for
  generator packages.
- It lets inference-side `laygen` remain Lightning-free.

Runner-up: `gentrain`. It is valid and short, but less consistent with the
existing `*gen` package family and reads more like a command than a shared
library. `genfit` is not selected because it over-emphasizes Lightning's
`trainer.fit` entrypoint while the library also owns parity, determinism, and
trace comparison utilities.

Proposed layout:

```text
lib/traingen/
  pyproject.toml
  src/traingen/
    __init__.py
    parity/
      __init__.py
      compare.py
      determinism.py
      trace.py
    lightning/
      __init__.py
      cli.py
      hooks.py
```

Dependency direction: model packages may depend on `traingen`; `traingen` may
depend on `laygen` only where layout schema helpers are needed; `laygen` must
not depend on `traingen`. Lightning dependencies should be isolated in
`traingen[lightning]` so normal inference imports do not pull in Lightning.

## Package Layout

```text
models/layout-flow/
  configs/
    training/
      layoutflow_rico25_vendor_compat.yaml
      layoutflow_publaynet_vendor_compat.yaml
      layoutflow_rico25_strict_parity.yaml
      layoutflow_publaynet_strict_parity.yaml
      smoke.yaml
  src/layout_flow/
    training/
      __init__.py
      cli.py
      config.py
      datamodule.py
      dataset.py
      lightning_module.py
      losses.py
      parity.py
      seed.py
```

Required symbols: `LayoutFlowLightningCLI`, `main`, `LayoutFlowTrainingModule`, `LayoutFlowDataModule`, `LayoutFlowH5Dataset`, `collate_layout_flow_batch`, and `LayoutFlowSeedMode`.

## CLI Contract

```bash
uv run --package layout-flow --extra training \
  python -m layout_flow.training.cli fit \
  --config models/layout-flow/configs/training/layoutflow_rico25_vendor_compat.yaml
```

```bash
CUDA_VISIBLE_DEVICES=0 CUBLAS_WORKSPACE_CONFIG=:4096:8 \
uv run --package layout-flow --extra training \
  python -m layout_flow.training.cli fit \
  --config models/layout-flow/configs/training/layoutflow_rico25_strict_parity.yaml \
  --trainer.limit_train_batches=2
```

`LayoutFlowLightningCLI` uses `subclass_mode_model=True` and `subclass_mode_data=True`, while package configs pin package-local defaults.

## LayoutFlow Vendor Mapping

Vendor `training_step` order is normative:

```text
cond_mask -> x0/x1 -> t -> xt/ut -> vt -> flow_loss + 0.2 * geom_l1_loss
```

Required trace points:

- batch tensors: `bbox`, `type`, `mask`, `length`, and stable sample id/HDF5 key where available
- `cond_mask`, `x0`, `x1`, `t`, `xt`, `ut`, `vt`
- `flow_loss`, `geom_l1_loss`, `train_loss`
- gradients, optimizer state, post-step parameter hashes/allclose, LR

LayoutFlow-specific values:

- PubLayNet: `num_labels=6`, `sampler_out_dim=7`
- RICO/RICO25: `num_labels=26`, `sampler_out_dim=9`
- default condition policy: `random4`
- default loss: MSE plus `geom_l1_loss` with weight `0.2`
- optimizer: AdamW, `lr=0.0005`, `betas=[0.9, 0.98]`
- scheduler: ReduceLROnPlateau on `FID_Layout`

Preserve vendor `random4` exactly:

```python
cond_mask = torch.ones((batch, seq, 4 + attr_dim), dtype=torch.int, device=device)
div = batch // 4
for i, length in enumerate(lengths[:div]):
    n = length * 0.2 * torch.rand(1, device=device)
    if length > 1:
        idx = torch.multinomial(torch.arange(length, device=device).float(), int(n.item() + 1))
        cond_mask[i, idx] = 0
cond_mask[div : 2 * div, :, 4:] = 0
cond_mask[2 * div : 3 * div, :, 2:] = 0
```

The final quarter remains all ones.

## Seed Modes

`vendor_compat`:

- mirrors vendor behavior as metadata/setup
- vendor imports call `torch.cuda.manual_seed_all(42975)`
- no `seed_everything`, no DataLoader generator, no worker seeding
- preserves vendor `torch.set_float32_matmul_precision("medium")` unless a parity harness explicitly records an override

`strict_deterministic`:

- uses `seed_everything: 42975`
- uses one selected GPU and `precision="32-true"`
- disables TF32 and cuDNN benchmark
- enables deterministic algorithms
- sets `CUBLAS_WORKSPACE_CONFIG` before process start
- captures/restores Python, NumPy, torch CPU, and torch CUDA RNG states
- uses `num_workers=0` for strict DataLoader stream comparison

## Dependencies

Shared `traingen[lightning]` extra:

```toml
[project.optional-dependencies]
lightning = [
    "lightning>=2.4,<3",
    "jsonargparse[signatures]>=4.27",
    "torchmetrics>=1.4",
]
```

The root or model `training` extra should depend on the workspace package
`traingen[lightning]` rather than making `laygen` own Lightning dependencies.

LayoutFlow model-local `training` extra:

```toml
training = [
    "traingen[lightning]",
    "h5pickle",
    "torchcfm",
    "torchdyn",
    "torchdiffeq",
]
```

`hydra-core` and `omegaconf` remain only in the vendor extra.

## First PR Scope

The first implementation PR should add `lib/traingen` skeleton, `layout_flow.training` skeleton, RICO/PubLayNet vendor-compatible YAMLs, strict parity YAMLs, smoke YAML, fast unit tests, and S0-S2 parity hooks/scripts. It should not attempt full validation/FID parity.

## Harness API

Generic harness utilities belong in `traingen`; LayoutFlow-specific adapters
remain local to `models/layout-flow/tests/vendor_parity` or
`layout_flow.training.parity`.

```python
@dataclass(frozen=True)
class TrainParitySpec:
    name: str
    vendor_factory: Callable[[], lightning.LightningModule]
    target_factory: Callable[[], lightning.LightningModule]
    batch_factory: Callable[[int], Mapping[str, torch.Tensor]]
    name_map: Mapping[str, str]
    trace_points: tuple[str, ...]
    tolerances: Mapping[str, TensorTolerance]
    determinism: DeterminismConfig

def trace_training_step(module, batch, rng_state, trace_points) -> StepTrace: ...
def compare_step_trace(vendor: StepTrace, target: StepTrace, tolerances) -> StepReport: ...
def compare_optimizer_step(vendor, target, tolerances) -> OptimizerStepReport: ...
def compare_batch_stream(vendor_loader, target_loader, *, steps: int) -> BatchStreamReport: ...
```

Candidate modules:

- `traingen.parity.determinism`: deterministic flags and RNG capture/restore.
- `traingen.parity.trace`: tensor summaries, hashes, first-mismatch reporting.
- `traingen.parity.compare`: tensor, state-dict, optimizer, and batch-stream
  comparators.
- `traingen.lightning`: manual training-step helpers, LR extraction, grad norms,
  shared CLI hooks, and optional base protocols/classes only after at least two
  models need the same abstraction.

## Tests And CI

Add root marker:

```toml
markers = [
    "integration: tests that require external downloads or heavyweight runtime setup",
    "training: tests that exercise Lightning training loops or training CLI entrypoints",
    "vendor_parity: requires original model weights and vendor dependencies",
]
```

Regular CI:

```bash
pytest -m "not vendor_parity and not integration and not training"
```

Unmarked tests cover config parsing, Hydra-key rejection, HDF5 collate conversion, `random4`, scalar `training_step`, and optimizer/scheduler wiring. Marked `training` tests run fast-dev CLI. Marked `vendor_parity` tests compare generated vendor traces and one optimizer step.

## Artifact Policy

Commit only metadata and regeneration instructions: config hashes, commands, revisions, environment, determinism flags, dataset/sample ids, seed/RNG hashes, per-step JSON summaries, tolerance rationale, and first-divergence reports. Do not commit golden tensors, checkpoints, downloaded datasets, generated images, or large intermediate artifacts.

## Conversion And Smoke

Training checkpoints must convert into the existing diffusers-style pipeline. Extend conversion helpers to accept new training checkpoint key prefixes, then run converted `LayoutFlowPipeline.from_pretrained` smoke tests.

```bash
uv run --package layout-flow python models/layout-flow/scripts/convert_original_checkpoint.py \
  --dataset rico25 \
  --checkpoint .cache/layout-flow/training-runs/rico25/checkpoints/<ckpt>.ckpt \
  --output-dir .cache/layout-flow/converted-trained/rico25

uv run --package layout-flow python models/layout-flow/scripts/smoke_from_pretrained.py \
  --path .cache/layout-flow/converted-trained/rico25
```

Implementation may rename or split the script if `convert_original_checkpoint.py` becomes misleading for newly trained checkpoints.

## Remaining Open Questions

- Whether first strict parity should set `num_sanity_val_steps=0` and `fid_calc_every_n=0`, or preserve vendor validation defaults and capture validation RNG before the first train step.
- Whether LayoutFlow should keep `torchcfm`, `torchdyn`, and `torchdiffeq` as model-local training deps or reuse already converted local flow/scheduler code where parity permits.
- Exact tolerance defaults after real Stage 1 LayoutFlow traces are measured.
- Exact issue labels and milestone names once the v0.5 umbrella meta and LayoutFlow training issue are created.

**Comment by creative-graphic-design-dev** [comment]
Owner decision amendment for the LayoutFlow training PR:

- Parity helpers must be split out of `traingen` into a separate workspace member, `lib/traingen-parity`, with import package `traingen_parity`.
- `lib/traingen` owns Lightning training primitives only, such as `traingen.lightning` CLI/hooks. It must not contain a `traingen.parity` subpackage.
- `lib/traingen-parity` owns deterministic runtime controls, step tracing, tensor comparisons, and related S0-S2 parity helpers via `traingen_parity.determinism`, `traingen_parity.trace`, and `traingen_parity.compare`.
- Dependency direction remains one-way: `traingen-parity` may depend on `traingen` only where needed, while `traingen` must not depend on `traingen-parity`; `laygen` depends on neither.
- LayoutFlow vendor-parity tests and training parity harnesses should import `traingen_parity`, not `traingen.parity`.
- Coverage remains enforced per workspace member, so both `traingen` and `traingen-parity` need independent 90%+ coverage evidence.

**Comment by creative-graphic-design-dev** [comment]
## Lesson: step-level parity (S0–S2) passing does not guarantee full-run (S5) parity

Recorded from the CGB-DM (#15 / #167) PKU vs CGL reproduction, because it affects how we set parity acceptance for every train-ourselves model.

**Observation.** PKU S0–S2 (per-step loss / gradient / post-Adam state, fixed-batch replay) passes within tight tolerances, and CGL full-run S5 is practical parity (~1–2σ). Yet PKU full-run S5 shows a systematic, one-directional divergence on the saliency/occlusion metrics: `unds` +0.019043 (~16σ pooled) and `occ` −0.010835 (~14σ pooled), while geometry/validity metrics match.

**What was ruled out (with evidence), and how.** The diagnosis order below is the reusable part:
1. Eval-side / data-input difference — ruled out: real PKU training inputs are bit-identical to vendor (`pixel_values`/`layout`/`saliency_box` max diff 0.0); same-sample eval scoring of the two checkpoints differs by ~1e-9 on `occ`; `unds` is geometry-only and never reads saliency.
2. Sampling stochasticity — ruled out: 3 S5 seeds are stable.
3. Therefore the gap is training-trajectory, not a data/eval/metric bug. The two independently trained models converge to different valid behaviors: ours places fewer but larger underlays (→ higher strict `unds` pass) and slightly smaller masks (→ lower `occ`). Step count matches (500×242).

**Still open.** Not yet separated between (a) benign training stochasticity and (b) a Lightning-vs-vendor training-loop orchestration difference. The one-directional ~16σ shift leans toward an orchestration difference over pure noise. Separating them requires a second full PKU training replicate with a different seed: run-to-run variance of similar magnitude ⇒ stochastic; a reproducible same-direction shift ⇒ orchestration difference (fixable).

## Protocol implications

- S0–S2 step parity is necessary but **not sufficient**. Always run S5 full-run parity per dataset; do not infer full-run parity from step parity.
- When S5 diverges, diagnose in this order before claiming a bug: (1) same-sample eval scoring of both checkpoints, (2) training-input bit-parity, (3) multi-seed S5 stability, (4) attribute the remainder to training trajectory.
- Parity thresholds should be per-dataset. Saliency/occlusion metrics (`unds`/`occ`) are trajectory-sensitive on PKU; a dataset can be practical-parity (CGL) while another is only qualitative with a documented caveat (PKU).
- To distinguish stochastic vs orchestration divergence, budget for at least one full training replicate with a different seed, or a per-epoch checkpoint curve comparison.

<details><summary>Raw PKU/CGL S5 deltas and distribution audit</summary>

- PKU S5 (package − reference): `unds` +0.019043 (16.4 pooled / 20.1 SEM σ), `occ` −0.010835 (−14.0 pooled / −17.2 SEM σ).
- CGL S5 (package − reference): `unds` +0.00410 (~1.84σ), `occ` +0.000484 (~1.62σ).
- Distribution audit (per 1000 PKU samples): ours valid elems 3364.7 vs 3742; underlays 621.7 vs 676; underlay area 0.0599 vs 0.0520; strict underlay pass 0.99143 vs 0.97239; mask area 0.12627 vs 0.12881.
- Vendor Epoch500 internal eval `unds` 0.978788 / `occ` 0.127215; external vendor S5 `unds` 0.972385 / `occ` 0.127496.
- Ours checkpoint `epoch=499-step=121000.ckpt` = 500×242 steps, matching vendor Epoch500 step count.

</details>

**Comment by creative-graphic-design-dev** [comment]
## Root cause: train-time RNG-stream orchestration divergence (PKU CGB-DM)

Follow-up to the earlier step-parity-vs-full-run-parity note. The CGB-DM PKU full-run S5 divergence (systematic +unds / -occ vs vendor, reproduced across two training seeds → not stochastic) has been root-caused, and the mechanism is general enough to affect other train-ourselves Lightning models.

**Isolation:** swapping samplers does not move the gap — ours weights under the vendor sampler keep the ours trend (unds 0.989 / occ 0.118), vendor weights under the ours sampler keep the vendor trend (unds 0.969 / occ 0.128). So the divergence is in the trained weights (training path), not eval/sampler/dataloader.

**Mechanism:** the vendor training loop runs in-loop validation sampling every 5 epochs after epoch 400 (`vendor/layout-dit/scripts/train_util.py:114`), and that sampling's reverse-DDIM consumes a large amount of the global torch RNG (`vendor/layout-dit/cgbdm/diffusion_utils.py:197`). The package Lightning recipe has no `validation_step` (log shows "Skipping val loop"), so from epoch 400 on the training timestep/noise RNG stream diverges from vendor. Quantified: PKU val 1000 samples × DDIM 100 steps ≈ 12.9M normal draws per validation, × 21 validations (epochs 400–500) ≈ 271M draws inserted into vendor's training RNG stream that the package run never makes. Different RNG stream → different final weights → the systematic metric offset, even though S0–S2 per-step computation is bit-identical.

## Protocol implication

- Full-run (S5) parity requires matching not just per-step math (S0–S2) but the **global RNG consumption schedule** of the vendor training loop. Anything the vendor does mid-training that draws from the global RNG (in-loop validation sampling, augmentation, EMA sampling, logging that samples) must be replicated in the package recipe, or the training trajectory diverges over a long run.
- Concretely: when porting a vendor training loop to Lightning, audit every global-RNG consumer in the vendor loop (especially periodic in-loop sampling/validation) and reproduce its RNG draws (a vendor-compatible validation/RNG callback), or seed/stream-isolate so the training RNG stream is not perturbed.
- This likely affects other train-ourselves Lightning ports with the same "no val-loop sampling" gap (DLT's smaller overlap/alignment offset is the prime suspect; a running DLT replicate will confirm).

Fix path for CGB-DM PKU: add a vendor-compatible train-time validation-sampling / RNG callback to the shared recipe so the RNG stream matches, then re-train PKU and re-run S5. Same fix likely needed for DLT and any similarly-ported model.
